ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes       #-}
+module Ouroboros.Storage.ImmutableDB.API
+  ( ImmutableDB(..)
+  , withDB
+  , getCurrentEpoch
+  , Iterator(..)
+
+  , IteratorResult(..)
+  , iteratorToList
+  , blobProducer
+
+  , module Ouroboros.Storage.ImmutableDB.Types
+  ) where
+
+import Control.Monad.Catch (MonadMask, bracket)
+import Control.Monad.Except (MonadError, ExceptT(ExceptT), runExceptT,
+                             throwError)
+
+import Data.Function (on)
+import Data.ByteString (ByteString)
+import Data.ByteString.Builder (Builder)
+
+import GHC.Stack (HasCallStack)
+
+import Pipes (Producer, yield, lift)
+
+import Ouroboros.Storage.FS.Class.Types
+import Ouroboros.Storage.ImmutableDB.Types
+
+
+-- | Open the database using the given function, perform the given action
+-- using the database, and closes the database using its 'closeDB' function,
+-- in case of success or when an exception was raised.
+withDB :: (HasCallStack, MonadMask m)
+       => m (Either ImmutableDBError (ImmutableDB m))
+          -- ^ How to open the database
+       -> (ImmutableDB m -> m (Either ImmutableDBError a))
+          -- ^ Action to perform using the database
+       -> m (Either ImmutableDBError a)
```

Comment:
IMO it would be better to have

```
withDB :: (MonadMask m)
       => m (ImmutableDB m)
          -- ^ How to open the database
       -> (ImmutableDB m -> m a)
          -- ^ Action to perform using the database
       -> m a
```

and let the particular `m` control how exceptions arise. For `IO`, real exceptions; for pure simulation, using `ExceptT`.

Reply:
Also, I'd even go as far as to get rid of the `MonadMask` and put this in `IO`, and make it particular to an `IO`-backed `ImmutableDB`. If it's a simulated DB, then there's no need to `bracket`.